File: skills/vector-memory/scripts/memory_flush.js
#!/usr/bin/env node
/**
 * Memory Flush Script (Node.js version matching Python API)
 * Flushes daily memory files into vector database
 * 
 * Usage: node memory_flush.js [--dry-run] [--force]
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { Pool } = require('pg');

// Config
const PG_CONFIG = {
    host: process.env.PGHOST || 'localhost',
    port: process.env.PGPORT || 5432,
    database: process.env.PGDATABASE || 'openclaw_memory',
    user: process.env.PGUSER || 'openclaw',
    password: process.env.PGPASSWORD || '',
};

const WORKSPACE = process.env.HOME + '/.openclaw/workspace';
const MEMORY_DIR = path.join(WORKSPACE, 'memory');
const FLUSH_TRACKER = path.join(MEMORY_DIR, 'vector-flush-tracker.json');
const GEMINI_KEY = process.env.GEMINI_API_KEY;
const EMBED_MODEL = 'gemini-embedding-001';

async function getEmbedding(text) {
    if (!GEMINI_KEY) {
        throw new Error('GEMINI_API_KEY environment variable not set');
    }
    
    const url = `https://generativelanguage.googleapis.com/v1beta/models/${EMBED_MODEL}:embedContent?key=${GEMINI_KEY}`;
    const payload = JSON.stringify({
        model: `models/${EMBED_MODEL}`,
        content: { parts: [{ text }] },
        outputDimensionality: 768
    });
    
    const response = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: payload
    });
    
    if (!response.ok) {
        throw new Error(`Gemini API error: ${response.status}`);
    }
    
    const data = await response.json();
    return data.embedding.values;
}

function loadTracker() {
    if (fs.existsSync(FLUSH_TRACKER)) {
        return JSON.parse(fs.readFileSync(FLUSH_TRACKER, 'utf-8'));
    }
    return { flushed_files: {} };
}

function saveTracker(tracker) {
    fs.writeFileSync(FLUSH_TRACKER, JSON.stringify(tracker, null, 2));
}

function fileHash(filepath) {
    const content = fs.readFileSync(filepath, 'utf-8');
    return crypto.createHash('md5').update(content).digest('hex');
}

function chunkMarkdown(text, sourceFile) {
    const chunks = [];
    const lines = text.split('\n');
    let currentSection = '';
    let currentText = [];
    
    for (const line of lines) {
        if (/^#{1,3}\s/.test(line)) {
            if (currentText.length > 0) {
                const content = currentText.join('\n').trim();
                if (content.length > 20) {
                    chunks.push({
                        text: content,
                        label: currentSection.replace(/^#+\s*/, '').trim(),
                        source_file: sourceFile
                    });
                }
            }
            currentSection = line;
            currentText = [line];
        } else {
            currentText.push(line);
        }
    }
    
    if (currentText.length > 0) {
        const content = currentText.join('\n').trim();
        if (content.length > 20) {
            chunks.push({
                text: content,
                label: currentSection.replace(/^#+\s*/, '').trim(),
                source_file: sourceFile
            });
        }
    }
    
    return chunks;
}

async function flush(dryRun = false, force = false) {
    const tracker = loadTracker();
    const pool = dryRun ? null : new Pool(PG_CONFIG);
    
    const files = fs.readdirSync(MEMORY_DIR)
        .filter(f => f.endsWith('.md'))
        .map(f => path.join(MEMORY_DIR, f));
    
    const memoryMd = path.join(WORKSPACE, 'MEMORY.md');
    if (fs.existsSync(memoryMd)) {
        files.push(memoryMd);
    }
    
    let totalStored = 0;
    
    for (const filepath of files) {
        const fname = path.basename(filepath);
        const fhash = fileHash(filepath);
        
        if (!force && tracker.flushed_files[fname] === fhash) {
            continue;
        }
        
        const content = fs.readFileSync(filepath, 'utf-8');
        const chunks = chunkMarkdown(content, fname);
        
        if (dryRun) {
            console.log(`[DRY RUN] ${fname}: ${chunks.length} chunks`);
            continue;
        }
        
        // Delete existing entries for this file
        await pool.query(
            "DELETE FROM memories WHERE metadata->>'source_file' = $1",
            [fname]
        );
        
        for (const chunk of chunks) {
            const embedding = await getEmbedding(chunk.text);
            const vecStr = '[' + embedding.join(',') + ']';
            
            await pool.query(
                `INSERT INTO memories (text, label, category, source, embedding, metadata)
                 VALUES ($1, $2, $3, $4, $5::vector, $6)`,
                [
                    chunk.text,
                    chunk.label,
                    'daily-note',
                    'flush',
                    vecStr,
                    JSON.stringify({ source_file: fname })
                ]
            );
            totalStored++;
        }
        
        tracker.flushed_files[fname] = fhash;
        console.log(`[FLUSHED] ${fname}: ${chunks.length} chunks stored`);
    }
    
    if (pool) await pool.end();
    saveTracker(tracker);
    
    console.log(JSON.stringify({ total_stored: totalStored, files_processed: files.length }));
}

// CLI usage
if (require.main === module) {
    const args = process.argv.slice(2);
    const dryRun = args.includes('--dry-run');
    const force = args.includes('--force');
    
    flush(dryRun, force)
        .then(() => process.exit(0))
        .catch(err => {
            console.error('Flush failed:', err.message);
            process.exit(1);
        });
}

module.exports = { flush, chunkMarkdown, fileHash };

File: skills/vector-memory/scripts/memory_flush.py
#!/usr/bin/env python3
"""Flush daily memory files into vector database."""
import argparse, glob, hashlib, json, os, re, sys, urllib.request
import psycopg2

DB = "host=localhost dbname=openclaw_memory user=openclaw password=openclaw123"
GEMINI_KEY = os.environ.get("GEMINI_API_KEY", "")
EMBED_MODEL = "gemini-embedding-001"
WORKSPACE = os.path.expanduser("~/.openclaw/workspace")
MEMORY_DIR = os.path.join(WORKSPACE, "memory")
FLUSH_TRACKER = os.path.join(MEMORY_DIR, "vector-flush-tracker.json")

def get_embedding(text):
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{EMBED_MODEL}:embedContent?key={GEMINI_KEY}"
    payload = json.dumps({"model": f"models/{EMBED_MODEL}", "content": {"parts": [{"text": text}]}, "outputDimensionality": 768}).encode()
    req = urllib.request.Request(url, data=payload, headers={"Content-Type": "application/json"})
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read())["embedding"]["values"]

def load_tracker():
    if os.path.exists(FLUSH_TRACKER):
        with open(FLUSH_TRACKER) as f:
            return json.load(f)
    return {"flushed_files": {}}

def save_tracker(t):
    with open(FLUSH_TRACKER, "w") as f:
        json.dump(t, f, indent=2)

def chunk_markdown(text, source_file):
    chunks, current_section, current_text = [], "", []
    for line in text.split("\n"):
        if re.match(r'^#{1,3}\s', line):
            if current_text:
                content = "\n".join(current_text).strip()
                if len(content) > 20:
                    chunks.append({"text": content, "label": current_section.strip("# ").strip(), "source_file": source_file})
            current_section, current_text = line, [line]
        else:
            current_text.append(line)
    if current_text:
        content = "\n".join(current_text).strip()
        if len(content) > 20:
            chunks.append({"text": content, "label": current_section.strip("# ").strip(), "source_file": source_file})
    return chunks

def file_hash(fp):
    with open(fp) as f:
        return hashlib.md5(f.read().encode()).hexdigest()

def flush(dry_run=False, force=False):
    tracker = load_tracker()
    conn = psycopg2.connect(DB) if not dry_run else None
    files = sorted(glob.glob(os.path.join(MEMORY_DIR, "*.md")))
    memory_md = os.path.join(WORKSPACE, "MEMORY.md")
    if os.path.exists(memory_md):
        files.append(memory_md)
    total_stored = 0
    
    for filepath in files:
        fname = os.path.basename(filepath)
        fhash = file_hash(filepath)
        if not force and fname in tracker["flushed_files"] and tracker["flushed_files"][fname] == fhash:
            continue
        with open(filepath) as f:
            content = f.read()
        chunks = chunk_markdown(content, fname)
        if dry_run:
            print(f"[DRY RUN] {fname}: {len(chunks)} chunks")
            continue
        cur = conn.cursor()
        cur.execute("DELETE FROM memories WHERE metadata->>'source_file' = %s", (fname,))
        for chunk in chunks:
            embedding = get_embedding(chunk["text"])
            vec_str = "[" + ",".join(str(v) for v in embedding) + "]"
            cur.execute("INSERT INTO memories (text, label, category, source, embedding, metadata) VALUES (%s,%s,%s,%s,%s::vector,%s)",
                        (chunk["text"], chunk["label"], "daily-note", "flush", vec_str, json.dumps({"source_file": fname})))
            total_stored += 1
        conn.commit()
        cur.close()
        tracker["flushed_files"][fname] = fhash
        print(f"[FLUSHED] {fname}: {len(chunks)} chunks stored")
    
    if conn:
        conn.close()
    save_tracker(tracker)
    print(json.dumps({"total_stored": total_stored, "files_processed": len(files)}))

if __name__ == "__main__":
    p = argparse.ArgumentParser()
    p.add_argument("--dry-run", action="store_true")
    p.add_argument("--force", action="store_true")
    args = p.parse_args()
    flush(args.dry_run, args.force)
